src/core/task/return.cc

```diff
@@ -199,6 +197,66 @@ void ReturnValues::legion_deserialize(const void* buffer)
   buffer_size_ = ptr - static_cast<const int8_t*>(buffer);
 }
 
+void ReturnValues::call_postamble(Legion::Context legion_context) const
+{
+  if (return_values_.empty()) {
+    Legion::Runtime::legion_task_postamble(legion_context);
+    return;
+  }
+
+#ifdef LEGATE_USE_CUDA
+  auto kind = Processor::get_executing_processor().kind();
+  if (kind == Processor::TOC_PROC) {
+    // FIXME: We don't currently have a good way to defer the return value packing on GPUs,
+    //        as doing so would require the packing to be chained up with all preceding kernels,
+    //        potentially launched with different streams, within the task. Until we find
+    //        the right approach, we simply synchronize the device before proceeding.
+    CHECK_CUDA(cudaDeviceSynchronize());
+
+    size_t return_size = legion_buffer_size();
+
+    auto mem_kind = (return_values_.size() == 1 && !has_exception_)
```

Comment:
We probably also want to skip this optimization if we're dealing with a task that has a single output region (and no other return values).